### PR TITLE
Add Foreign Table References to wandb.Table

### PR DESCRIPTION
**Pull request by tssweeney**
Description
-----------

As part of the larger release plan for DSViz/Dartasets (https://paper.dropbox.com/doc/Artifacts-Media-DSviz-APIs-a6NEEU63S9CZvL4iTpKha), we want to give users the ability to easily associate tables. Not only is this sort of functionality key to a great experience in the Keras Callback, but is allow fundamental to enable easy-to-declare, yet flexible table association. Currently, the only way to associate tables to to construct an explicit "join table" which is a hardcoded mapping from one table to another. There are a number of problems with this approach: it is static, users have to learn a new object type, it unable to properly validate inputs, and it results in awkward code. This PR does the following (ordered from highest level to lowest level):

1. The `ndx` produced from the row iterator automatically links tables. So the following code will automatically link `new_table` to `train_table`. This is stored in the table metadata and can be acted on in the UI. Moreover, when download `new_table` from an artifact, `train_table` is available and can be accessed (we will make this easier in the future).
```
for ndx, row in train_table.iterrows():
    new_table.add_data(ndx, predict(row))
```
2. In a similar vein, it is common for people to have their own id system, so we enable specifying a primary key for this reason. Let's assume `train_table` has columns 'id', 'image', 'label'. Then the following code will automatically create a reference from `new_table` to `train_table`. That is because row[0] is a special object that knows where it came from, but acts like a string. (I had a previous version that changes `ndx` to this value if a PK is set, but i think that is a bit confusing as the return type of iterrows is dependent on if PK is set.
```
train_table.set_pk('id')
for ndx, row in train_table.iterrows():
    new_table.add_data(row[0], predict(row[1]))
```
So at this point, we have enabled automatic table linking for the vast majority of use cases. Ok, going one level deeper:
3. If the user already has tables filled out, they can make an explicit call to set a column as a foreign key reference. The following code does the same thing as the above code, but provides a slightly different api which is suitable for different circumstances.:
```
new_table.set_fk('fk_col', train_table, 'id')
```
4. Ok, now you are a rogue developer looking to do something gnarly. you can access the lowest level which is direct column casting. Note: this API is not expected to be used by users, but are the building blocks of the above.
```
# cast a column as a pk
table.cast('id', wandb.data_types._TablePrimaryKeyType())
# cast a column as a fk
new_table.cast('fk_col', wandb.data_types._TableForeignKeyType(table, 'id'))
# cast a column as a foreign index
new_table.cast('fk_col', wandb.data_types._TableForeignIndexType(table))
```

Note: since the meat of the tracking implementation is implemented in the type system, there are not new json attributes, resulting in no need for json schema management/migrations. Moreover, all of the serialization and deserialization properties which create asset references when the asset is already in another artifact are preserved. This means in the above examples, if `train_table` has not been logged, then it will be logged to the same artifact as `new_table`. Yet, if it came from another artifact, then a reference is stored instead. Yay!

Testing
-------

How was this PR tested? New Unit tests!

### CODE REVIEW

File: wandb/data_types.py [hunk 1]
```diff
@@ -290,25 +344,46 @@ def add_data(self, *data):
                     len(self.columns), self.columns
                 )
             )
-        self._validate_data(data)
+        data = list(data)
```
Comment (shawnlewis) [inline wandb/data_types.py:347]: Do you need to convert to a list here? enumerate should work on any iterable.

File: tests/test_tables.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    assert [id_ for id_, row in list(table.iterrows())] == [0, 1]
+
+    # Cast as a PK
+    if use_helper:
+        table.set_pk("id")
+    else:
+        table.cast("id", wandb.data_types._TablePrimaryKeyType())
+
+    assert all(
+        [row[0]._table == table and row[0]._col_name == "id" for row in table.data]
```
Comment (shawnlewis) [inline tests/test_tables.py:40]: Ideally we wouldn't access underscore prefix methods like this, even in tests. (test the interface, not the implementation). Just a general comment, don't worry about it in this PR.

File: wandb/data_types.py [hunk 3]
```diff
@@ -411,12 +487,121 @@ def iterrows(self):
...
+
+            self._pk_col = _pk_col
+            self._fk_cols = _fk_cols
+
+        # Apply updates to data only if there are update or the caller
+        # requested the final row to be updated
+        if has_update or force_last:
+            self._apply_key_updates(not has_update)
+
+    def _apply_key_updates(self, only_last=False):
```
Comment (shawnlewis) [inline wandb/data_types.py:557]: Style / design: a boolean switch argument that completely changes what a method does can usually be written better as two separate methods.

**Review by shawnlewis — APPROVED** [review]
This is awesome

File: tests/test_tables.py (hunk 2 above)
Comment (tssweeney) [reply tests/test_tables.py:40]: That is fair, I should have made accessors on the special classes to get these without underscores. Noted for the future.

File: wandb/data_types.py (hunk 1 above)
Comment (tssweeney) [reply wandb/data_types.py:347]: since the *data parameter wraps the parameters in a tuple, i think it is needed :/

Comment (tssweeney) [reply wandb/data_types.py]: fixed

**Pull request merged**